### ISSUE TITLE
conda install creates `ValueError: Incompatible component merge:`

### ISSUE THREAD

**Issue by kkiesling** [conda/conda#11442]
### Checklist

- [X] I added a descriptive title
- [X] I searched for other issues and couldn't find a solution or duplication
- [X] I already searched in Google and didn't find any good information or help
- [X] I looked at the [docs](https://docs.conda.io/projects/conda/en/latest/user-guide/index.html) and didn't see anything to help

### What happened?

I believe this is a `conda install` error at this point, not a conda-forge issue despite being a conda-forge package, which is why I am making this issue here. I have also searched/asked on the conda-forge gitter with no solution, have googled, and searched the conda, conda-forge, and conda-build repos/orgs on github and I don't see this issue reported anywhere yet.

We have recently updated a conda-forge package called [DAGMC](https://github.com/conda-forge/dagmc-feedstock) to include MPI builds. It relies on a package called [MOAB](https://github.com/conda-forge/moab-feedstock) which supplies [this as the `run_exports`](https://github.com/conda-forge/moab-feedstock/blob/main/recipe/meta.yaml#L45) with `mpi` and `tempest` variables in the string. After updating DAGMC on conda-forge, conda-forge CI passes and building DAGMC locally (`conda build recipe/ -c conda-forge`) succeeds just fine. However, when then trying to install the package with `conda install -c conda-forge dagmc`, it fails due to this "Incompatible component merge" shown below. Because of the `tempest` in the error, we believe it is connected to the MOAB dependency, but we have not figured out in MOAB how to fix this either and believe at this point it is beyond either of these two packages. Furthermore, the package can be installed with `mamba` successfully in the same environment with `mamba install -c conda-forge dagmc`.

Expected behavior is that `conda install -c conda-forge dagmc` is successful.

```
>> conda install -c conda-forge dagmc
Collecting package metadata (current_repodata.json): failed

# >>>>>>>>>>>>>>>>>>>>>> ERROR REPORT <<<<<<<<<<<<<<<<<<<<<<

    Traceback (most recent call last):
      File "/home/kkiesling/miniconda3/lib/python3.9/site-packages/conda/exceptions.py", line 1082, in __call__
        return func(*args, **kwargs)
      File "/home/kkiesling/miniconda3/lib/python3.9/site-packages/conda/cli/main.py", line 87, in _main
        exit_code = do_call(args, p)
      File "/home/kkiesling/miniconda3/lib/python3.9/site-packages/conda/cli/conda_argparse.py", line 84, in do_call
        return getattr(module, func_name)(args, parser)
      File "/home/kkiesling/miniconda3/lib/python3.9/site-packages/conda/cli/main_install.py", line 20, in execute
        install(args, parser, 'install')
      File "/home/kkiesling/miniconda3/lib/python3.9/site-packages/conda/cli/install.py", line 260, in install
        unlink_link_transaction = solver.solve_for_transaction(
      File "/home/kkiesling/miniconda3/lib/python3.9/site-packages/conda/core/solve.py", line 152, in solve_for_transaction
        unlink_precs, link_precs = self.solve_for_diff(update_modifier, deps_modifier,
      File "/home/kkiesling/miniconda3/lib/python3.9/site-packages/conda/core/solve.py", line 195, in solve_for_diff
        final_precs = self.solve_final_state(update_modifier, deps_modifier, prune, ignore_pinned,
      File "/home/kkiesling/miniconda3/lib/python3.9/site-packages/conda/core/solve.py", line 300, in solve_final_state
        ssc = self._collect_all_metadata(ssc)
      File "/home/kkiesling/miniconda3/lib/python3.9/site-packages/conda/common/io.py", line 88, in decorated
        return f(*args, **kwds)
      File "/home/kkiesling/miniconda3/lib/python3.9/site-packages/conda/core/solve.py", line 463, in _collect_all_metadata
        index, r = self._prepare(prepared_specs)
      File "/home/kkiesling/miniconda3/lib/python3.9/site-packages/conda/core/solve.py", line 1058, in _prepare
        reduced_index = get_reduced_index(self.prefix, self.channels,
      File "/home/kkiesling/miniconda3/lib/python3.9/site-packages/conda/core/index.py", line 291, in get_reduced_index
        push_record(record)
      File "/home/kkiesling/miniconda3/lib/python3.9/site-packages/conda/core/index.py", line 265, in push_record
        combined_depends = record.combined_depends
      File "/home/kkiesling/miniconda3/lib/python3.9/site-packages/conda/models/records.py", line 340, in combined_depends
        result = {ms.name: ms for ms in MatchSpec.merge(self.depends)}
      File "/home/kkiesling/miniconda3/lib/python3.9/site-packages/conda/models/match_spec.py", line 478, in merge
        reduce(lambda x, y: x._merge(y, union), group) if len(group) > 1 else group[0]
      File "/home/kkiesling/miniconda3/lib/python3.9/site-packages/conda/models/match_spec.py", line 478, in <lambda>
        reduce(lambda x, y: x._merge(y, union), group) if len(group) > 1 else group[0]
      File "/home/kkiesling/miniconda3/lib/python3.9/site-packages/conda/models/match_spec.py", line 509, in _merge
        final = this_component.merge(that_component)
      File "/home/kkiesling/miniconda3/lib/python3.9/site-packages/conda/models/match_spec.py", line 760, in merge
        raise ValueError("Incompatible component merge:\n  - %r\n  - %r"
    ValueError: Incompatible component merge:
      - 'mpi_openmpi_*'
      - 'mpi_openmpi_tempest*'
```

### Conda info

```shell
>> conda info

     active environment : debug
    active env location : /home/kkiesling/miniconda3/envs/debug
            shell level : 2
       user config file : /home/kkiesling/.condarc
 populated config files : 
          conda version : 4.12.0
    conda-build version : 3.21.8
         python version : 3.9.7.final.0
       virtual packages : __cuda=11.4=0
                          __linux=5.4.0=0
                          __glibc=2.31=0
                          __unix=0=0
                          __archspec=1=x86_64
       base environment : /home/kkiesling/miniconda3  (writable)
      conda av data dir : /home/kkiesling/miniconda3/etc/conda
  conda av metadata url : None
           channel URLs : https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /home/kkiesling/miniconda3/pkgs
                          /home/kkiesling/.conda/pkgs
       envs directories : /home/kkiesling/miniconda3/envs
                          /home/kkiesling/.conda/envs
               platform : linux-64
             user-agent : conda/4.12.0 requests/2.27.1 CPython/3.9.7 Linux/5.4.0-94-generic ubuntu/20.04.4 glibc/2.31
                UID:GID : 1012:1016
             netrc file : None
           offline mode : False

```
```


### Conda config

```shell
This command resulted in no output
```

### Conda list

```shell
>> conda list --show-channel-urls
# packages in environment at /home/kkiesling/miniconda3/envs/debug:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       2_gnu    conda-forge
bzip2                     1.0.8                h7f98852_4    conda-forge
c-ares                    1.18.1               h7f98852_0    conda-forge
ca-certificates           2021.10.8            ha878542_0    conda-forge
curl                      7.83.0               h2283fc2_0    conda-forge
hdf4                      4.2.15               h10796ff_3    conda-forge
hdf5                      1.12.1          nompi_h4df4325_104    conda-forge
jpeg                      9e                   h166bdaf_1    conda-forge
keyutils                  1.6.1                h166bdaf_0    conda-forge
krb5                      1.19.3               h08a2579_0    conda-forge
ld_impl_linux-64          2.36.1               hea4e1c9_2    conda-forge
libblas                   3.9.0           14_linux64_openblas    conda-forge
libcblas                  3.9.0           14_linux64_openblas    conda-forge
libcurl                   7.83.0               h2283fc2_0    conda-forge
libedit                   3.1.20191231         he28a2e2_2    conda-forge
libev                     4.33                 h516909a_1    conda-forge
libffi                    3.4.2                h7f98852_5    conda-forge
libgcc-ng                 11.2.0              h1d223b6_16    conda-forge
libgfortran-ng            11.2.0              h69a702a_16    conda-forge
libgfortran5              11.2.0              h5c6108e_16    conda-forge
libgomp                   11.2.0              h1d223b6_16    conda-forge
liblapack                 3.9.0           14_linux64_openblas    conda-forge
libnetcdf                 4.8.1           nompi_h329d8a1_102    conda-forge
libnghttp2                1.47.0               he49606f_0    conda-forge
libnsl                    2.0.0                h7f98852_0    conda-forge
libopenblas               0.3.20          pthreads_h78a6416_0    conda-forge
libssh2                   1.10.0               ha35d2d1_2    conda-forge
libstdcxx-ng              11.2.0              he4da1e4_16    conda-forge
libuuid                   2.32.1            h7f98852_1000    conda-forge
libzip                    1.8.0                h1c5bbd1_1    conda-forge
libzlib                   1.2.11            h166bdaf_1014    conda-forge
metis                     5.1.0             h58526e2_1006    conda-forge
moab                      5.3.1           nompi_tempest_py310h6a0b23b_5    conda-forge
ncurses                   6.3                  h27087fc_1    conda-forge
numpy                     1.22.3          py310h4ef5377_2    conda-forge
openssl                   3.0.2                h166bdaf_1    conda-forge
pip                       22.0.4             pyhd8ed1ab_0    conda-forge
pymoab                    5.3.1                    pypi_0    pypi
python                    3.10.4          h2660328_0_cpython    conda-forge
python_abi                3.10                    2_cp310    conda-forge
readline                  8.1                  h46c0cb4_0    conda-forge
setuptools                62.1.0          py310hff52083_0    conda-forge
sqlite                    3.38.2               h4ff8645_0    conda-forge
tempest-remap             2.1.1                hfda0864_0    conda-forge
tk                        8.6.12               h27826a3_0    conda-forge
tzdata                    2022a                h191b570_0    conda-forge
wheel                     0.37.1             pyhd8ed1ab_0    conda-forge
xz                        5.2.5                h516909a_1    conda-forge
zlib                      1.2.11            h166bdaf_1014    conda-forge
```
```


### Additional Context

DAGMC feedstock repo: https://github.com/conda-forge/dagmc-feedstock
MOAB feedstock repo: https://github.com/conda-forge/moab-feedstock

**Comment by kkiesling** [comment]
Also possibly relevant, this same failure occurs on Linux and MacOS (I have tried both).

**Comment by acaprez** [comment]
We're also seeing this issue trying to build what are (to the best of my knowledge) completely unrelated packages - `cp2k` and `lammps`.  I've tried adding both `moab` and `dagmc` to the `disallowed_packages` list to no effect.

**Comment by vsisl** [comment]
I'm facing the same problem. [Here](https://github.com/pyne/pyne/issues/1452) are more details about my setup.

**Comment by pawel125** [comment]
I've got very similar issue here: [#4498](https://github.com/conda/conda-build/issues/4498). Did you solve it?

**Comment by pawel125** [comment]
@kkiesling I can build the evironment with all the packages I would like to use, but I cannot build my own package - then the error occurs. Hopefully someone will notice my issue.
Thanks for advice!

**Comment by snajder-r** [comment]
@pawel125 I have the same issue running `conda-build`, but I think it's not actually the building that's the problem. It's the tests (which run `conda install` to create a test environment)
I found as a temporary workaround I can run `conda-build` with the `--no-test` parameter. Then it works. My users might still experience the error when trying to install the package, but they can then use `mamba install` as a workaround.

**Comment by pawel125** [comment]
@snajder-r Unfortunately it did not work for me. I have noticed that the error occurs because of the bioconductor-genomeinfodb, bioconductor-plyranges and r-vcfr dependencies. Building countinues beyond the point of the error when I drop those dependencies, unfortunatelly I need them in my package.

**Comment by skupr-anaconda** [comment]
It seems like this PR on conda-forge tries to fix that issue https://github.com/conda-forge/dagmc-feedstock/pull/20

**Comment by kkiesling** [comment]
@skupr-anaconda that PR was made by me (also the creator of this issue). That PR did not/does not fix the issue.

**Comment by Jiayuan-Gu** [comment]
I encounter a similar issue, then I find it is because I set `channel_priority: strict` in `~/.condarc`. If I remove that option, the issue disappears. Hope it can provide some information to fix the bug.

**Comment by kkiesling** [comment]
> I encounter a similar issue, then I find it is because I set `channel_priority: strict` in `~/.condarc`. If I remove that option, the issue disappears. Hope it can provide some information to fix the bug.

I don't have this file and a `grep` for `channel_priority:` on my system indicates I don't have that set anywhere.

**Comment by LukeLabrie** [comment]
Seeing the same issue

**Comment by kkiesling** [comment]
@LukeLabrie thanks for the explanation.

We have already confirmed that DAGMC, MOAB, etc all follow the same MPI implementations/variants. This issue is also not specific to me. Every person who has tried to install DAGMC since implementing MPI variants has received the same error.

**Comment by kkiesling** [comment]
The error also indicates that it is the same MPI variant:

```
ValueError: Incompatible component merge:
      - 'mpi_openmpi_*'
      - 'mpi_openmpi_tempest*'
```

**Comment by minrk** [comment]
To be clear, this _is_ a conda bug, where [GlobStrMatch](https://github.com/conda/conda/blob/86447b6bc47eb9549db9c857fef941e2111cb7a4/conda/models/match_spec.py#L813) doesn't override the base [MatchInterface.merge](https://github.com/conda/conda/blob/86447b6bc47eb9549db9c857fef941e2111cb7a4/conda/models/match_spec.py#L756), so only exact matches are allowed. GlobStrMatch should instead compute the _intersecting_ pattern, rather than require exact matches.

**Comment by minrk** [comment]
#11612 fixes the issue with tests. I can't be sure if there will be any negative consequences when there are multiple incompatible globs attempted, which would previously (correctly) fail immediately, but will now fail later in the solve. For instance:

```
python -m conda install petsc=*=complex_* petsc=*=real_*
```

eventually fails with:

```
PackagesNotFoundError: The following packages are not available from current channels:

  - petsc[build='^(?=complex_.*)(?:real_.*)$']
```

which seems correct, though isn't represented as a conflict anymore, but rather a single unsatisfiable dependency.

**Comment by dprada** [comment]
Same error here using `conda build` when `--python 3.10` (working well with 3.9 or 3.8):

```
  ValueError: Incompatible component merge:
    - 'mpi_mpich_*'
    - 'mpi_mpich_tempest*'
  Error: Process completed with exit code 1.
```

**Comment by h-vetinari** [comment]
This is breaking the installation of seemingly unrelated packages around abseil etc.: https://github.com/conda-forge/abseil-cpp-feedstock/issues/46

**Comment by ChrisBarker-NOAA** [comment]
I'm having the same issue as @dprada :

Trying to build a package of pyopengl-accelerate for python 3.10 -- it works for 3.9

ValueError: Incompatible component merge:
    - 'mpi_mpich_*'
    - 'mpi_mpich_tempest*'
  Error: Process completed with exit code 1.

It seems to be when setting up the environment for building -- not anything I've specified.

**Comment by wilderfield** [comment]
I landed here for tangential reason, but was also seeing this error when trying to do conda-build for my own custom conda package. As others have suggested, removing channel priority strict seems to fix the problem. 
I don't fully understand the problem, and I'm not adding much value here... but remember to try this if you face this issue...
Check ~/.condarc

**Comment by arodel21** [comment]
I am having the same issue. I tried removing the channel priority strict and python 3.8 instead of the version 3.10. Nothing seems to fix it, except for using other channels instead of conda-forge, but one of my packages is cairosvg and it seems like there is no other channel with the latest version of this module. Does anyone have more updates on this matter?

**Comment by ekomarova** [comment]
Hi, @minrk! Could you please tell me if it is planned to fix this problem in the near future by merging this PR https://github.com/conda/conda/pull/11612?

**Comment by minrk** [comment]
@ekomarova I'm not a conda maintainer, so I can't comment on plans. My understanding is that #11612 is ready to go and fixes this problem. It is only waiting on maintainer approval.

**Comment by jehturner** [comment]
I kept running into this error unpredictably at the `TEST` stage of `conda-build` on `osx-64`, in between other build problems, but seem to have found the answer, pending an upstream fix.

I don't really understand the nature of this "component merge" check -- I'm not even using MPI and adding a bit of debugging to `conda` seemed to indicate that it was processing irrelevant dependency lists when the failure happened. By process of elimination, however, **I was able to resolve the problem by going into the `conda-bld/` subdirectory, deleting the package cache with something like `rm -fr osx-64/.cache osx-64/*repodata*` and then doing `conda index .`**. I don't believe I've seen the problem again since doing that. Creating a new build installation should therefore also solve it.

I think what led to this may have been building the same packages from a mixture of Anaconda and (later) `conda-forge` channels (possibly without changing the build number during some troubleshooting), which somehow left the cache in a conflicted state? Anyway, HTH.

**Comment by slava92** [comment]
Thanks! I was able to build a package after wiping out "conda-build:" directory.

**Comment by JanWielemaker** [comment]
Got the same mpich/mpi_mpich_ with a fresh install of miniconda on Linux after adding conda-forge.  For me it worked by removing the `defaults` channel from ~/.condarc (i.e., only keeping `conda-forge`)

**Comment by jezdez** [comment]
Sorry for reassigning this again, but we had other higher priorities on our plate, but this will definitely happen at some point.

**Comment by skupr-anaconda** [comment]
I'm not 100% sure and it really can be the issue with conda-build's cache.

I got this error again. I guess I found the root cause. 
1. I built dependencies of the package `X` and all of them were uploaded to the private channel `test`. So conda-build tries to build the package `X `using dependencies from the private channel `test`.
2. Then I'm building the **same** dependencies _locally_ (it's **process 1**) and I think two identical dependencies with the same name and versions conda-build tries to pick up (one from cache and one from the conda-bld/<platform_name> folder). 
3. But at the same time, I'm building a lot of packages in parallel (it's the **process 2**; `X` among them). So two conda-build processes are running in different consoles and conda-build wants to pick up dependencies from the `test` channel and `from` local. And race condition happens.

In short, the issue can be somewhere in between the local and remote channels or two channels AND caches

**Comment by h-vetinari** [comment]
> Sorry for reassigning this again, but we had other higher priorities on our plate, but this will definitely happen at some point.

Half a year later, are there any plans/ideas when that "at some point" might be?
